### ISSUE TITLE
targetLang="en" is deprecated, please use "en-GB" or "en-US" instead

### ISSUE THREAD

**Issue by kalebheitzman** [DeepLcom/deepl-php#44]
```
Execute scheduled task: Autotranslation Fetch (filter_autotranslate\task\autotranslate_task)
... started 22:06:02. Current memory use 29.1 MB.
Debugging increased temporarily due to faildelay of 60
Executing autotranslation fetch jobs...
200 jobs found...
fetching en translation for d41d8cd98f00b204e9800998ecf8427e key...
... used 4 dbqueries
... used 0.015782117843628 seconds
Scheduled task failed: Autotranslation Fetch (filter_autotranslate\task\autotranslate_task),targetLang="en" is deprecated, please use "en-GB" or "en-US" instead.
Backtrace:
* line 156 of /filter/autotranslate/vendor/deeplcom/deepl-php/src/Translator.php: call to DeepL\Translator->buildBodyParams()
* line 108 of /filter/autotranslate/classes/task/autotranslate_task.php: call to DeepL\Translator->translateText()
* line 405 of /lib/classes/cron.php: call to filter_autotranslate\task\autotranslate_task->execute()
* line 208 of /lib/classes/cron.php: call to core\cron::run_inner_scheduled_task()
* line 125 of /lib/classes/cron.php: call to core\cron::run_scheduled_tasks()
* line 186 of /admin/cli/cron.php: call to core\cron::run_main_process()
```

This isn't technically a bug but it broke our entire app. The Readme might need updated to notify users of this. We have a custom plugin on Moodle that's based entirely on 2 letter locale codes and I haven't gotten into the internals yet to see if it will give us the en-US format. There might be other software out there that suffers from the same problem.

**Comment by JanEbbing** [comment]
Hi, what is the bug here exactly? The information is in the readme:

```
Some target languages also include the regional variant according to ISO 3166-1, for example 'en-US', or 'pt-BR'. The source language also accepts null, to enable auto-detection of the source language.
```

and this is not a recent change, we haven't accepted `en` as a target language for years.

**Comment by kalebheitzman** [comment]
@JanEbbing that's weird. Our plugin was working fine when using just **en** as a target language code and it just broke in the last few weeks with that deprecation notice. The documentation states that you have to pass a locale code in order for the glossary to work and unless we change our Moodle install to use regional variants we'll have to hardcode locales into the plugin to make it work with DeepL now. 

On the readme it shows ```$translationResult = $translator->translateText('Hello, world!', 'en', 'fr');``` but my understanding is that I now have to use en-US or en-GB, etc and that I cannot use en. Am I misunderstanding something here?

**Comment by JanEbbing** [comment]
Hi @kalebheitzman - not sure why it broke, but we did not change anything in this regard. DeepL differentiates between a `sourceLanguage` (language the text is in, can be null in order to be inferred) and a `targetLanguage` (language to translate the text into) - the `targetLanguage` needs the regional variant to be specified, else it will throw an error (unfortunately only once you actually call the function, as we want users of older library versions to be able to use languages released after the library version).

In the example

`$translationResult = $translator->translateText('Hello, world!', 'en', 'fr');`

we call `translate` with a `sourceLanguage` of `'en'` and a `targetLanguage` of `'fr'`, which is fine. On the other hand,

`$translationResult = $translator->translateText('Bonjour, le monde!', 'fr', 'en');`

would throw an error, you would need to write something like

`$translationResult = $translator->translateText('Bonjour, le monde!', 'fr', 'en-GB');`

**Comment by nlemoine** [comment]
@JanEbbing I stumbled on that issue too in an app that basically provide a source and target selectors the user can select from.

The list of those selectors are hydrated from DeepL `/languages` endpoint. What's confusing here is that we expect (as developers) those lists items to be valid when using them and not waiting to run a translation process to throw errors. Sadly, there's no way to identify the deprecated codes (e.g. `en` and `pt`) and I have to keep a target language blacklist that I would prefer be maintained from your side. 

One way to solve this would be to introduce a `is_deprecated` (or whatever name) data on the returned languages byt your API (just like `supports_formality`) which could be then mapped to a prop on the `DeepL\Language` class. That would solve developer's problems with this.

What do you think?

**Comment by JanEbbing** [comment]
Hi @nlemoine , please note that the `/languages` endpoint has a `type` parameter that can be passed to specify if you want to retrieve source or target languages, which defaults to source, hence you would get `en` and `pt`, as they are valid source languages. If I set `type` to `target`, I correctly get the regional variants (`EN-US`, `PT-BR`, etc). Does this solve your issue around a blacklist?

**Comment by jlherren** [comment]
I was also bitten by this when switching from manual curl requests to using this library. The API itself still fully supports `en` and `pt` as target languages. It's just this library that doesn't.

Useful to know when encountering this issue: I empirically found that the API treats `en` as `en-US` and `pt` as `pt-PT`.

**Comment by JanEbbing** [comment]
Yes, the API needed to be backwards compatible, but as it just decides a default variant we decided to not support this in client libraries, which came later, to ensure highest translation quality for customers.

**Comment by JanEbbing** [comment]
Hi, when you create a glossary the source and target languages need to be what we call "base langauges", languages without variants (eg. "EN"), while target languages for translation need to be with variants. Note that glossaries can be used for translation regardless of variant, e.g. a glossary with source language "DE" and target language "EN" can be used to translate from "DE" into "EN-GB" and from "DE" into "EN-US".

There is a utility method `removeRegionalVariant` in `LanguageCode.php` in this package you can use to potentially simplify your code (strip regional variants before creating glossaries).
